### DISCUSSION TITLE
Using XRANK and Sort - Can it be done gracefully?

### DISCUSSION THREAD

**Discussion by StephenHoughton4** [microsoft-search/pnp-modern-search#3106] [Q&A]
Hi,

I previously had a Ticket relating to using XRANK #2985 , which I found to be great for Weighting Results, but I hgave a question regarding Ordering.

Is it possible to use XRANK and Sort? I used "Rank" in the Sort By Column, but using Debug mode I saw the Rank Numbers were all unique.

Example I have 7 users:
John Lee, Lee Burgess, Leeroy King, Adam Lee, Lee Jones, Neil Leeson, Dani Leeson

Using the XRANK formula:
`{?(((((
...
)
XRANK(cb=1600) PersonInitials:({searchTerms})) 
XRANK(cb=800) LastName:({searchTerms})) 
XRANK(cb=400) FirstName:({searchTerms}))  
XRANK(cb=200) LastName:({searchTerms}*)) 
XRANK(cb=100) FirstName:({searchTerms}*))  `

So my requirements for the results are:
- I expect to See the order of Surname "Lee", First Name "Lee", LastName Starts with "Lee" FirstName starts with Lee.
- I expect the Order to be by Surname, then First Name (After the Rank Boosts.

**Example Searching * (With Rank not in Sort Settings):**
Results:
Lee Burgess, Lee Jones, Adam Lee, John Lee, Neil Leeson, Dani Leeson, Leeroy King
_This Works As Expected (Searching *)_

**Example Searching Lee (Rank being used in Sort By)**
Results:
John Lee, Adam Lee, Lee Burgess, Lee Jones, Neil Leeson, Dani Leeson, Leeroy King
_So Surname Lee is first, but the First names are not in alphabetical order (Same with the Surnames being Leeson). Note this is an example, so the order when I include Location Leeds the latter results are all over the place. It is not just related to First name / surname_

Preferred (Expected):
Adam Lee, John Lee, Lee Burgess, Lee Jones, Dani Leeson, Neil Leeson, Leeroy King
_Ordered by Surname Lee (then FirstName A-Z), FirstName Lee (then by LastName A-Z), Surname starts Lee (then by LastName A-z, followed by FirstName A-z) etc..._

Is here a way to make the PNP Search produce results in an order like this? THe Debug mode shows each result has a random Rank value (which is higher on the ones matching XRANK, but still all unique values).

Thanks
Stephen H

**Comment by wobba** [comment]
Using cb in rank is shifty these days. Better to use stdb. And you cannot have sorting in the KQL in addition to XRANK. Better to fix this inside the handlebar template maybe if all is in the resultset.

Or try something like the below assuming the property you want to sort on is refinable or sortable.

![image](https://github.com/microsoft-search/pnp-modern-search/assets/51104/3d6160d3-998b-4674-aac3-5edda7a190fc)

**Reply by StephenHoughton4** [reply]
@wobba Thanks for the reply, I was using stdb originally, changed it to cb to see if it helped Rank at all..

As for the Sort by Rank then LastName, every result has a a rank of 6 decimal places, which means the Rank number is unique, so doesn't matter what is added to the 2nd sort field it seems. :(

Thanks

**Reply by wobba** [reply]
Tricky indeed. I think fetching a large enough result set for page 1 and resorting in the template is your best option.
